src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -106,35 +106,21 @@ private static unsafe SafeFileHandle CreateFile(string fullPath, FileMode mode,
             return fileHandle;
         }
 
-        private static unsafe void Preallocate(string fullPath, long preallocationSize, SafeFileHandle fileHandle)
+        private static unsafe void Preallocate(string fullPath, long preallocationSize, SafeFileHandle fileHandle, FileMode mode)
         {
-            var allocationInfo = new Interop.Kernel32.FILE_ALLOCATION_INFO
+            // preallocationSize must be ignored for non-seekable files, unsupported file systems
+            // and other failures other than ERROR_DISK_FULL and ERROR_FILE_TOO_LARGE
+            if (!FileStreamHelpers.TrySetFileLength(fileHandle, preallocationSize, out int errorCode)
+                && errorCode == Interop.Errors.ERROR_DISK_FULL || errorCode == Interop.Errors.ERROR_FILE_TOO_LARGE)
             {
-                AllocationSize = preallocationSize
-            };
-
-            if (!Interop.Kernel32.SetFileInformationByHandle(
-                fileHandle,
-                Interop.Kernel32.FileAllocationInfo,
-                &allocationInfo,
-                (uint)sizeof(Interop.Kernel32.FILE_ALLOCATION_INFO)))
-            {
-                int errorCode = Marshal.GetLastPInvokeError();
-
                 // we try to mimic the atomic NtCreateFile here:
                 // if preallocation fails, close the handle and delete the file
                 fileHandle.Dispose();
                 Interop.Kernel32.DeleteFile(fullPath);
```

Comment:
Can you remind me why we're trying to mimic the NtCreateFile behavior here?

Reply:
> Can you remind me why we're trying to mimic the NtCreateFile behavior here?

When `posix_fallocate` fails on Linux because there is not enough space, it usually consumes all available disk space (so the system is in a very bad shape).
Since we want to avoid that and we also do want to have the same behavior on every OS, we just remove the file (no matter if it's empty (Windows) or incomplete (Linux)) and throw an exception.

Reply:
Ok, in that case isn't the comment misleading?  We're not doing this because of NtCreateFile behavior as the comment states, rather we're doing this because we want to have the same behavior as on unix and on unix we believe we need to delete.

Reply:
good point, I am going to update the comment